service/s3/endpoints.go

```diff
@@ -213,6 +215,127 @@ func finalizeClientEndpointResolverOptions(options *Options) {
 
 }
 
+type legacyEndpointResolverAdapter struct {
+	legacyResolver EndpointResolver
+	resolver       EndpointResolverV2
+}
+
+func (l *legacyEndpointResolverAdapter) ResolveEndpoint(ctx context.Context, params EndpointParameters) (endpoint smithyendpoints.Endpoint, err error) {
+	var fips aws.FIPSEndpointState
+	var dualStack aws.DualStackEndpointState
+
+	if aws.ToBool(params.UseFIPS) {
+		fips = aws.FIPSEndpointStateEnabled
+	}
+	if aws.ToBool(params.UseDualStack) {
+		dualStack = aws.DualStackEndpointStateEnabled
+	}
+
+	resolveEndpoint, err := l.legacyResolver.ResolveEndpoint(aws.ToString(params.Region), EndpointResolverOptions{
+		ResolvedRegion:       aws.ToString(params.Region),
+		UseFIPSEndpoint:      fips,
+		UseDualStackEndpoint: dualStack,
+	})
+	if err != nil {
+		return endpoint, err
+	}
+
+	if resolveEndpoint.HostnameImmutable {
+		uriString := resolveEndpoint.URL
+		uri, err := url.Parse(uriString)
+		if err != nil {
+			return endpoint, fmt.Errorf("Failed to parse uri: %s", uriString)
+		}
+
+		return smithyendpoints.Endpoint{
+			URI: *uri,
+		}, nil
+	}
+
+	if resolveEndpoint.Source == aws.EndpointSourceServiceMetadata {
+		return l.resolver.ResolveEndpoint(ctx, params)
+	}
+
+	params.UseDualStack = nil
+	params.UseFIPS = nil
+	params.Endpoint = &resolveEndpoint.URL
+
+	return l.resolver.ResolveEndpoint(ctx, params)
+}
+
+type isDefaultProvidedImplementation interface {
+	isDefaultProvidedImplementation()
+}
+
+type compatibleEndpointResolver struct {
+	EndpointResolverV2 EndpointResolverV2
+}
+
+func (n *compatibleEndpointResolver) isDefaultProvidedImplementation() {}
+
+func (n *compatibleEndpointResolver) ResolveEndpoint(region string, options EndpointResolverOptions) (endpoint aws.Endpoint, err error) {
+	reg := region
+	fips := options.UseFIPSEndpoint
+	if len(options.ResolvedRegion) > 0 {
+		reg = options.ResolvedRegion
+	} else {
+		reg, fips = mapPseudoRegion(region)
+	}
+	ctx := context.Background()
+	resolved, err := n.EndpointResolverV2.ResolveEndpoint(ctx, EndpointParameters{
+		Region:       &reg,
+		UseFIPS:      aws.Bool(fips == aws.FIPSEndpointStateEnabled),
+		UseDualStack: aws.Bool(options.UseDualStackEndpoint == aws.DualStackEndpointStateEnabled),
+	})
+	if err != nil {
+		return endpoint, err
+	}
+
+	endpoint = aws.Endpoint{
+		URL:               resolved.URI.String(),
+		HostnameImmutable: false,
+		Source:            aws.EndpointSourceServiceMetadata,
+	}
+
+	return endpoint, nil
+}
+
+func mapPseudoRegion(pr string) (region string, fips aws.FIPSEndpointState) {
```

Comment:
**suggestion** This could use some documentation on why it exists (I'm _assuming_ because the legacy resolver allowed pseudo regions but the new resolver doesn't need them so we have to map them to the appropriate region + param).

Reply:
gotcha, will add documentation. and yah your assumption seems to be correct
https://github.com/aws/aws-sdk-go-v2/blob/main/service/s3/endpoints.go#L183